internal/rsl/rsl.go

```diff
@@ -338,66 +390,175 @@ func GetFirstEntry(repo *git.Repository) (*Entry, error) {
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	if !knowsCommit {
-		return nil, ErrNoRecordOfCommit
+		return nil, nil, ErrNoRecordOfCommit
 	}
 
 	for {
-		iteratorEntry, err := GetNonGittufParentForEntry(repo, firstEntry)
+		iteratorEntry, iteratorAnnotations, err := GetNonGittufParentForEntry(repo, firstEntry)
 		if err != nil {
 			if errors.Is(err, ErrRSLEntryNotFound) {
-				return firstEntry, nil
+				return firstEntry, firstAnnotations, nil
 			}
-			return nil, err
+			return nil, nil, err
 		}
 
 		knowsCommit, err := gitinterface.KnowsCommit(repo, iteratorEntry.TargetID, commit)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
-
 		if !knowsCommit {
-			return firstEntry, nil
+			return firstEntry, firstAnnotations, nil
 		}
 
 		firstEntry = iteratorEntry
+		firstAnnotations = iteratorAnnotations
+	}
+}
+
+// GetEntriesInRange returns a list of standard entries between the specified
+// range and a map of annotations that refer to each standard entry in the
+// range. The annotations map is keyed by the ID of the standard entry, with the
+// value being a list of annotations that apply to that standard entry.
+func GetEntriesInRange(repo *git.Repository, firstID, lastID plumbing.Hash) ([]*Entry, map[plumbing.Hash][]*Annotation, error) {
+	return GetEntriesInRangeForRef(repo, firstID, lastID, "")
+}
+
+// GetEntriesInRangeForRef returns a list of standard entries for the ref
+// between the specified range and a map of annotations that refer to each
+// standard entry in the range. The annotations map is keyed by the ID of the
+// standard entry, with the value being a list of annotations that apply to that
+// standard entry.
+func GetEntriesInRangeForRef(repo *git.Repository, firstID, lastID plumbing.Hash, refName string) ([]*Entry, map[plumbing.Hash][]*Annotation, error) {
+	// We have to iterate from latest to get the annotations that refer to the
+	// last requested entry
+	iterator, err := GetLatestEntry(repo)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	allAnnotations := []*Annotation{}
+	for iterator.GetID() != lastID {
+		// Until we find the entry corresponding to lastID, we just store
+		// annotations
+		if annotation, isAnnotation := iterator.(*Annotation); isAnnotation {
+			allAnnotations = append(allAnnotations, annotation)
+		}
+
+		parent, err := GetParentForEntry(repo, iterator)
+		if err != nil {
+			return nil, nil, err
+		}
+		iterator = parent
+	}
+
+	entryStack := []*Entry{}
+	inRange := map[plumbing.Hash]bool{}
+	for iterator.GetID() != firstID {
+		// Here, all items are relevant until the one corresponding to first is
+		// found
+		switch it := iterator.(type) {
+		case *Entry:
+			if len(refName) != 0 && it.RefName != refName && !strings.HasPrefix(it.RefName, GittufNamespacePrefix) {
+				// If refName is set AND the iterator's ref is not for the
+				// specified ref AND the iterator's ref is not for a gittuf
+				// namespace, break
+				break
+			}
```

Comment:
It looks like we want to skip these, but IIUC if we encounter this we'll break out and add it to the entry stack anyway on lines 530-533. Is this intentional or do we need to do a similar check below?

Reply:
The `break` here moves execution to line 520. https://go.dev/ref/spec#Break_statements

So, line 530 is encountered only at the end of the for loop to handle the first entry. That said, there _is_ a bug because the same check isn't done there to see if it's a relevant entry.

I'm going to do two things. First, I'm going to invert this check to get rid of the break statement. Second, I'll add the same conditions at 530.

Reply:
Done!